### DISCUSSION TITLE
OpenGL backend

### DISCUSSION THREAD

**Discussion by ghost** [fltk-rs/fltk-rs#1428] [Q&A] [answered]
@MoAlyousef on your blog post announcing 1.0 you write

> And the OpenGL backend was extended to allow drawing widgets using OpenGL. That means GlWindow can now display widgets, if you need hardware acceleration or need to display widgets on top of 3D graphics!

Does that mean there's a backend that, unlike Wayland backend using Cairo, is using OpenGL to draw widgets and windows?

I'm not sure how to interpret the comments in `fltk/Cargo.toml` and what you wrote in the blog post

```toml
enable-glwindow = ["fltk-sys/enable-glwindow"] # Support for systems without OpenGL
legacy-opengl   = ["fltk-sys/legacy-opengl"]   # (Experimental) Support of Lagacy OpenGL
```

**Comment by MoAlyousef** [comment, marked as answer]
> Does that mean there's a backend that, unlike Wayland backend using Cairo, is using OpenGL to draw widgets and windows?

Yes for the widgets. The window will use whatever backend you're using.
```
 cargo run --example temp --features=enable-glwindow,use-wayland
```
So here it'll use a wayland window (surface). Without it it will use an XWindow. On windows it'll be an HWND and macos an NSWindow.

The minimal example would be:
```rust
use fltk::{enums::*, prelude::*, *};

fn main() {
    let a = app::App::default();
    let mut wind = window::GlWindow::new(100, 100, 400, 300, None);
    wind.set_mode(Mode::Opengl3|Mode::MultiSample);
    wind.set_frame(FrameType::FlatBox);
    wind.set_color(Color::Background);
    wind.begin();
    let btn = button::Button::new(160, 200, 80, 30, "Click");
    wind.end();
    wind.show();
    a.run().unwrap();
}
```
Notice the explicit call to Window::begin() which will notify the GlWindow to accept children. We also set a frame_type and a color for it to appear like a normal FLTK window.
The set_mode enables OpenGL version 3 or above, and multisample enables antialiasing.

**Reply by ghost** [reply]
Do some of the themes draw without OpenGL, or is this all abstracted away?

**Reply by ghost** [reply]
For the record, as with all OpenGL windows (on Wayland), there's a base memory overhead of ~100MiB which is required regardless of what you draw. Just a note for those with a use case where this is significant.

**Reply by ghost** [reply]
The (possibly) infinite memory growth ("leak") when repeatedly zooming in/out the FLTK window, which I had reported a few months ago, seems to be significantly smaller with the current Wayland backend. Before, growth was happening in large strides and never settling.

**Reply by ghost** [reply]
Can I set up GlWindow conditionally and expose this in my app behind a feature flag, while everything else in the app would be unaware of window type? Besides explicit OpenGL FLTK API use, I suppose.

**Reply by MoAlyousef** [reply]
> Do some of the themes draw without OpenGL, or is this all abstracted away?

It should all be abstracted away.
Note: I haven't personally extensively used it yet. It's supposed to be an experimental feature of FLTK itself, so there may be bugs or other issues.

> Can I set up GlWindow conditionally and expose this in my app behind a feature flag, while everything else in the app would stay the same? Besides explicit OpenGL FLTK API use, I suppose.

Yes. You can use a add a feature flag to your crate:
```toml
# Cargo.toml
[features]
use-glwin = ["fltk/enable-glwindow"]
```
In your source code:
```rust
use fltk::{enums::*, prelude::*, *};

#[cfg(feature = "use-glwin")]
fn create_win() -> window::GlWindow {
    let mut wind = window::GlWindow::new(100, 100, 400, 300, None);
    wind.set_mode(Mode::Opengl3|Mode::MultiSample);
    wind.set_frame(FrameType::FlatBox);
    wind.set_color(Color::Background);
    wind.begin();
    wind
}

#[cfg(not(feature = "use-glwin"))]
fn create_win() -> window::Window {
    window::Window::new(100, 100, 400, 300, None)
}

fn main() {
    let a = app::App::default();
    let mut wind = create_win();
    let btn = button::Button::new(160, 200, 80, 30, "Click");
    wind.end();
    wind.show();
    a.run().unwrap();
}
```
